File: 1365.有多少小于当前数字的数字.js
/*
 * @lc app=leetcode.cn id=1365 lang=javascript
 *
 * [1365] 有多少小于当前数字的数字
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var smallerNumbersThanCurrent = function(nums) {
    let cnt = new Array(101).fill(0);
    let n = nums.length;

    for (let i = 0; i < n; i++) {
        cnt[nums[i]]++;
    }

    for (let i = 1; i <= 100; i++) {
        cnt[i] += cnt[i - 1];
    }

    let res = [];
    for (let i = 0; i < n; i++) {
        res[i] = nums[i] === 0 ? 0 : cnt[nums[i] - 1];
    }

    return res;
};
// @lc code=end

File: 1365.有多少小于当前数字的数字.java
/*
 * @lc app=leetcode.cn id=1365 lang=java
 *
 * [1365] 有多少小于当前数字的数字
 */

// @lc code=start
class Solution {
    public int[] smallerNumbersThanCurrent(int[] nums) {
        // 可以想到用数组来标记出现过的数字,然后对每个数字统计它前面有多少个小数字

        // 数字范围[0,100]总共 101 个
        int[] cnt = new int[101];
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            cnt[nums[i]]++;
        }

        // 累加每个数字前有多少个小数字
        // 这里累加后是小于等于当前数字的数量
        for (int i = 1; i <= 100; i++) {
            cnt[i] += cnt[i - 1];
        }
        int[] ret = new int[n];
        for (int i = 0; i < n; i++) {
            // cnt前一个是小于自己的数量
            ret[i] = nums[i] == 0 ? 0 : cnt[nums[i] - 1];
        }
        return ret;
    }
}
// @lc code=end

